percona/controller/pgcluster/controller_test.go

```diff
@@ -226,3 +233,216 @@ func havePMMSidecar(sts appsv1.StatefulSet) bool {
+
+func getReconcileCount(r *postgrescluster.Reconciler) int {
+	return r.Tracer.(*tracerWithCounter).counter
+}
+
+var _ = Describe("Watching secrets", Ordered, func() {
+	ctx := context.Background()
+
+	const crName = "watch-secret-test"
+	const ns = crName
+
+	crunchyR := crunchyReconciler()
+	r := reconciler()
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      crName,
+			Namespace: ns,
+		},
+	}
+
+	mgrCtx, cancel := context.WithCancel(ctx)
+	wg := sync.WaitGroup{}
+
+	BeforeAll(func() {
+		By("Creating the Namespace to perform the tests")
+		err := k8sClient.Create(ctx, namespace)
+
+		Expect(err).To(Not(HaveOccurred()))
+		mgr, err := runtime.CreateRuntimeManager(ns, cfg, true)
+		Expect(err).To(Succeed())
+		Expect(v2beta1.AddToScheme(mgr.GetScheme())).To(Succeed())
+
+		r.Client = mgr.GetClient()
+		crunchyR.Client = mgr.GetClient()
+		crunchyR.Tracer = &tracerWithCounter{t: crunchyR.Tracer}
+
+		cm := &perconaController.CustomManager{Manager: mgr}
+		Expect(crunchyR.SetupWithManager(cm)).To(Succeed())
+
+		Expect(cm.Controller()).NotTo(BeNil())
+		r.CrunchyController = cm.Controller()
+		Expect(r.SetupWithManager(mgr)).To(Succeed())
+
+		wg.Add(1)
+		go func() {
+			Expect(mgr.Start(mgrCtx)).To(Succeed())
+			wg.Done()
+		}()
+	})
+
+	AfterAll(func() {
+		By("Stopping manager")
+		cancel()
+		wg.Wait()
+
+		By("Deleting the Namespace to perform the tests")
+		_ = k8sClient.Delete(ctx, namespace)
+	})
+
+	cr, err := readDefaultCR(crName, ns)
+	It("should read default cr.yaml", func() {
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	reconcileCount := 0
+	Context("Create cluster and wait until Reconcile stops", func() {
+		It("should create PerconaPGCluster and PostgresCluster", func() {
+			Expect(k8sClient.Create(ctx, cr)).Should(Succeed())
+
+			Eventually(func() error {
+				return k8sClient.Get(ctx, client.ObjectKeyFromObject(cr), new(v2beta1.PerconaPGCluster))
+			}).Should(BeNil())
+
+			Eventually(func() error {
+				return k8sClient.Get(ctx, client.ObjectKeyFromObject(cr), new(v1beta1.PostgresCluster))
+			}).Should(BeNil())
+		})
+
+		It("should wait until PostgresCluster will stop to Reconcile multiple times", func() {
+			Eventually(func() bool {
+				pgCluster := new(v1beta1.PostgresCluster)
+				err := k8sClient.Get(ctx, client.ObjectKeyFromObject(cr), pgCluster)
+				if err != nil {
+					return false
+				}
+				// When ManagedFields get field with `status` subresource, crunchy's Reconcile stops being called
+				for _, f := range pgCluster.ManagedFields {
+					if f.Manager == postgrescluster.ControllerName && f.Subresource == "status" {
+						return true
+					}
+				}
+
+				return false
+			}, time.Second*30, time.Millisecond*250).Should(Equal(true))
+			reconcileCount = getReconcileCount(crunchyR)
+		})
+
+		It("should reconcile 0 times", func() {
+			Eventually(func() int { return getReconcileCount(crunchyR) }, time.Second*15, time.Millisecond*250).
+				Should(Equal(reconcileCount))
```

Comment:
Should you explicitly check here that `reconcileCount` is zero, like `Expect(reconcileCount).Should(BeZero())`?

What if `reconcileCount` and `getReconcileCount(crunchyR)` are equal but it is something other than zero?

And maybe for all the tests, I think you can remove `reconcileCount` and assert if `getReconcileCount(crunchyR)` is the exact number you expect, rather than stating that number in the `It` block description and then assuming that `reconcileCount` is that same number.

Two reasons for this:
- We can potentially have false positives;
- It is more clear when you read the exact thing in the `Expect` assertion.

Reply:
We can't remove `reconcileCount` and use exact number. In the `should wait until PostgresCluster will stop to Reconcile multiple times` container, we can get different results from `getReconcileCount` (from 2 to 4 during my tests). However, after [this](https://github.com/percona/percona-postgresql-operator/blob/d5a1f491ca6312da10b8a9059ef8dd4217fc979b/percona/controller/pgcluster/controller_test.go#L336-L341) condition is met, crunchy's `Reconcile` will stop being called automatically. That's why I treat `reconcileCount` as 0 here.

We can't use `Expect` for such checks because we have to wait until the Reconcile is called after updating secrets. But I agree that in that specific place, the check is redundant since it will be always positive

Reply:
https://github.com/percona/percona-postgresql-operator/pull/443/commits/ee1b3fcd3363a29374c7bc44009d24597ed2a512